README.md
## aTalk (Jabber / XMPP)
- an encrypted instant messaging with video call and GPS features for android

<p align="center">
    <a href="https://cmeng-git.github.io/atalk">
        <img src="./art/atalk.png" alt="aTalk">
    </a>
    &nbsp;
    <a href="https://cmeng-git.github.io/atalk/faq.html">
        <img src="./art/help.png" alt="FAQ">
    </a>
    &nbsp;
    <a href="https://play.google.com/store/apps/details?id=org.atalk.ohos&hl=en">
        <img src="./art/google_play.png" alt="Google PlayStore">
    </a>
    &nbsp;
    <a href="https://f-droid.org/en/packages/org.atalk.ohos/">
        <img src="./art/fdroid-logo.png" alt="F-Droid">
    </a>
    &nbsp;
    <a href="https://www.youtube.com/watch?v=9w5WwphzgBc">
        <img src="./art/youtube.png" alt="YouTube">
    </a>
</p>


## Features
aTalk is an xmpp client designed for android and supports the following features:
* Instant messaging in plain text and End-to-End encryption with [OMEMO](https://conversations.im/omemo/) or [OTR](https://otr.cypherpunks.ca/)
* SSL Certificate authentication, DNSSEC and DANE Security implementation for enhanced secure Connection Establishment
* OMEMO encryption in group chat session enhancing privacy and security
* OMEMO Media File Sharing for all files including Stickers, Bitmoji and Emoji rich contents
* Support http file upload for file sharing with offline contact and in group chat
* Support Stickers, Bitmoji and Emoji rich content sharing via Google Gboard
* Send and receive files for all document types and images with thumbnail preview and gif animation
* Auto accept file transfer with max file size option
* Implement fault-tolerance file transfer algorithm, via Jingle File Transfer, Legacy In-Band/SOCK5 Bytestreams and HttpFileUpload protocols to ease and enhance file sharing reliability
* Enhance and harmonize UI for file sharing in chat and chatRoom
* Support share, quote and forward of messages and media with previews prior to sending
* Support multi-instances of audio media player with functions: Play, Pause, Resume, Stop and Seek
* Support unread message badges in contact and chatRoom list views
* User selectable option for heads-up notification and quite hours
* Share of social media links are tagged with thumbnail and title
* Support Text to Speech and speech recognition UI with enable option per chat session (contact and group chat)
* XEP-0012: Last Activity time associated with contacts
* XEP-0048: Bookmarks for conference room and autoJoin on login
* XEP-0070: Verifying HTTP Requests via XMPP entity for user authentication without password entry
* XEP-0071: XHTML-IM Support chat messages containing lightweight text markup
* XEP-0085: Chat State Notifications
* XEP-0124: Bidirectional-streams Over Synchronous HTTP (BOSH) with Proxy support
* XEP-0178: Use of SASL EXTERNAL with TLS Certificates for client authentication
* XEP-0184: Message Delivery Receipts with user enable/disable option
* XEP-0215: External Service Discovery
* XEP-0251: Jingle Session Transfer: Support both Unattended and Attended call transfer
* XEP-0313: Message Archive Management
* XEP-0391: Jingle Encrypted Transports for OMEMO encrypted media file sharing
* Implement call waiting to accept a second incoming call by placing the in-progress call on hold; and allow switching between calls
* Implement Jabber VoIP-PBX gateway Telephony, allowing PBX phone call via service gateway
* Join or create room with full room configuration support for owner
* Integrated captcha protected room user interface with retry on failure
* Support both voice and video call with ZRTP, SDES and DTLS SRTP encryption modes
* ZRTP crypto algorithms uses SHA-2 384, 256bit ciphers AES256 & TWOFISH-256, enabling elliptic with Curve 25519
* SDES encryption with AES256 and AES192, acceleration using OpenSSL
* Support simultaneous media call and message chat sessions
* Unique GPS-Location implementation as standalone tool, send locations to your desired buddy for real-time tracking or playback animation
* A 360° street view of your current location use for self-guided tour. The street view tracks and follows your direction of sight
* Built-in demo for GPS-Location features
* Integrated photo editor with zooming and cropping, user can update the avatar with ease
* Last message correction, message carbons and offline messages (OMEMO)
* Stream Management and mobile network ping interval self-tune optimization support
* In-Band Registration with captcha option support
* Multiple accounts creation
* User selectable Themes support
* Multi-language UI support (Bahasa Indonesia, Chinese Simplify, English, German, Portuguese, Russian, Slovak and Spanish)

## XMPP Standards Implemented

aTalk works seamlessly with almost every XMPP servers available on network, limited only by servers features supported.<br/>
It supports the following XEP's, standards for XMPP clients.

* [XEP-0012: Last Activity 2.0](https://xmpp.org/extensions/xep-0012.html)
* [XEP-0030: Service Discovery 2.5rc3](https://xmpp.org/extensions/xep-0030.html)
* [XEP-0045: Multi-User Chat 1.34.3](https://xmpp.org/extensions/xep-0045.html)
* [XEP-0047: In-Band Bytestreams 2.0.1](https://xmpp.org/extensions/xep-0047.html)
* [XEP-0048: Bookmarks 1.2](https://xmpp.org/extensions/xep-0048.html)
* [XEP-0054: vcard-temp 1.2](https://xmpp.org/extensions/xep-0054.html)
* [XEP-0060: Publish-Subscribe 1.24.1](https://xmpp.org/extensions/xep-0060.html)
* [XEP-0065: SOCKS5 Bytestreams 1.8.2](https://xmpp.org/extensions/xep-0065.html)
* [XEP-0070: Verifying HTTP Requests via XMPP 1.0.1](https://xmpp.org/extensions/xep-0070.html)
* [XEP-0071: XHTML-IM 1.5.4](https://xmpp.org/extensions/xep-0071.html)
* [XEP-0077: In-Band Registration 2.4](https://xmpp.org/extensions/xep-0077.html)
* [XEP-0084: User Avatar 1.1.4](https://xmpp.org/extensions/xep-0084.html)
* [XEP-0085: Chat State Notifications 2.1](https://xmpp.org/extensions/xep-0085.html)
* [XEP-0092: Software Version 1.1](https://xmpp.org/extensions/xep-0092.html)
* [XEP-0095: Stream Initiation 1.2](https://xmpp.org/extensions/xep-0095.html)
* [XEP-0096: SI File Transfer 1.3.1](https://xmpp.org/extensions/xep-0096.html)
* [XEP-0100: Gateway Interaction 1.0](https://xmpp.org/extensions/xep-0100.html)
* [XEP-0115: Entity Capabilities 1.6.0](https://xmpp.org/extensions/xep-0115.html)
* [XEP-0124: Bidirectional-streams Over Synchronous HTTP (BOSH) 1.11.2](https://xmpp.org/extensions/xep-0124.html)
* [XEP-0138: Stream Compression 2.1](https://xmpp.org/extensions/xep-0138.html)
* [XEP-0153: vCard-Based Avatar 1.1](https://xmpp.org/extensions/xep-0153.html)
* [XEP-0158: CAPTCHA Forms 1.5.8](https://xmpp.org/extensions/xep-0158.html)
* [XEP-0163: Personal Eventing Protocol 1.2.2](https://xmpp.org/extensions/xep-0163.html)
* [XEP-0166: Jingle 1.1.2](https://xmpp.org/extensions/xep-0166.html)
* [XEP-0167: Jingle RTP Sessions 1.2.1](https://xmpp.org/extensions/xep-0167.html)
* [XEP-0172: User Nickname 1.1](https://xmpp.org/extensions/xep-0172.html)
* [XEP-0176: Jingle ICE-UDP Transport Method 1.1.1](https://xmpp.org/extensions/xep-0176.html)
* [XEP-0177: Jingle Raw UDP Transport Method 1.1.1](https://xmpp.org/extensions/xep-0177.html)
* [XEP-0178: Best Practices for Use of SASL EXTERNAL with Certificates 1.2](https://xmpp.org/extensions/xep-0178.html)
* [XEP-0184: Message Delivery Receipts 1.4.0](https://xmpp.org/extensions/xep-0184.html)
* [XEP-0198: Stream Management 1.6](https://xmpp.org/extensions/xep-0198.html)
* [XEP-0199: XMPP Ping 2.0.1](https://xmpp.org/extensions/xep-0199.html)
* [XEP-0203: Delayed Delivery 2.0](https://xmpp.org/extensions/xep-0203.html)
* [XEP-0206: XMPP Over BOSH 1.4](https://xmpp.org/extensions/xep-0206.html)
* [XEP-0215: External Service Discovery 1.0.0](https://xmpp.org/extensions/xep-0215.html)
* [XEP-0231: Bits of Binary 1.0](https://xmpp.org/extensions/xep-0231.html)
* [XEP-0234: Jingle File Transfer 0.19.1](https://xmpp.org/extensions/xep-0234.html)
* [XEP-0237: Roster Versioning 1.3](https://xmpp.org/extensions/xep-0237.html)
* [XEP-0249: Direct MUC Invitations 1.2](https://xmpp.org/extensions/xep-0249.html)
* [XEP-0251: Jingle Session Transfer 0.2](https://xmpp.org/extensions/xep-0251.html)
* [XEP-0260: Jingle SOCKS5 Bytestreams Transport Method 1.0.3](https://xmpp.org/extensions/xep-0260.html)
* [XEP-0261: Jingle In-Band Bytestreams Transport Method 1.0](https://xmpp.org/extensions/xep-0261.html)
* [XEP-0262: Use of ZRTP in Jingle RTP Sessions 1.0](https://xmpp.org/extensions/xep-0262.html)
* [XEP-0264: File Transfer Thumbnails 0.4](https://xmpp.org/extensions/xep-0264.html)
* [XEP-0278: Jingle Relay Nodes 0.4.1](https://xmpp.org/extensions/xep-0278.html)
* [XEP-0280: Message Carbons 1.0.1](https://xmpp.org/extensions/xep-0280.html)
* [XEP-0293: Jingle RTP Feedback Negotiation 1.0.1](https://xmpp.org/extensions/xep-0293.html)
* [XEP-0294: Jingle RTP Header Extensions Negotiation 1.1.1](https://xmpp.org/extensions/xep-0294.html)
* [XEP-0298: Delivering Conference Information to Jingle Participants (Coin) 0.2](https://xmpp.org/extensions/xep-0298.html)
* [XEP-0308: Last Message Correction 1.2.0](https://xmpp.org/extensions/xep-0308.html)
* [XEP-0313: Message Archive Management 1.0.1](https://xmpp.org/extensions/xep-0313.html)
* [XEP-0319: Last User Interaction in Presence 1.0.2](https://xmpp.org/extensions/xep-0319.html)
* [XEP-0320: Use of DTLS-SRTP in Jingle Sessions 1.0.0](https://xmpp.org/extensions/xep-0320.html)
* [XEP-0338: Jingle Grouping Framework 1.0.0](https://xmpp.org/extensions/xep-0338.html)
* [XEP-0339: Source-Specific Media Attributes in Jingle 1.0.1](https://xmpp.org/extensions/xep-0339.html)
* [XEP-0343: Signaling WebRTC datachannels in Jingle 0.3.1](https://xmpp.org/extensions/xep-0343.html)
* [XEP-0352: Client State Indication 1.0.0](https://xmpp.org/extensions/xep-0352.html)
* [XEP-0353: Jingle Message Initiation 0.4.0](https://xmpp.org/extensions/xep-0353.html)
* [XEP-0363: HTTP File Upload 1.1.0](https://xmpp.org/extensions/xep-0363.html)
* [XEP-0364: Off-the-Record Messaging (V2/3) 0.3.2](https://xmpp.org/extensions/xep-0364.html)
* [XEP-0371: Jingle ICE Transport Method 0.3.1](https://xmpp.org/extensions/xep-0371.html)
* [XEP-0384: OMEMO Encryption 0.8.3](https://xmpp.org/extensions/xep-0384.html)
* [XEP-0391: Jingle Encrypted Transports 0.1.2](https://xmpp.org/extensions/xep-0391.html)
* [XEP-0441: Message Archive Management Preferences 0.2.0](https://xmpp.org/extensions/xep-0441.htmll)
* [XEP-xxxx: OMEMO Media sharing 0.0.2](https://xmpp.org/extensions/inbox/omemo-media-sharing.html)

## Acknowledgements

Libraries used in this project:

* [Android Support Library](https://developer.android.com/topic/libraries/support-library/index.html)
* [android-betterpickers](https://github.com/code-troopers/android-betterpickers)
* [Android-EasyLocation](https://github.com/akhgupta/Android-EasyLocation)
* [android-youtube-player](https://github.com/PierfrancescoSoffritti/android-youtube-player)
* [annotations-java5](https://mvnrepository.com/artifact/org.jetbrains/annotations)
* [apache-mime4j-core](https://github.com/apache/james-mime4j)
* [bouncycastle](https://github.com/bcgit/bc-java)
* [butterknife](https://github.com/JakeWharton/butterknife)
* [ckChangeLog](https://github.com/cketti/ckChangeLog)
* [commons-text](https://commons.apache.org/proper/commons-text/)

aTalk/ReleaseNotes.txt
Project aTalk-Android Release Notes
=========================================================================
Version:        00.02.010 (0002010)
Release Date:   12/29/2023
Author:         cmeng
- Update aTalk-ohos for android API-34 support (needs master jarjar-gradle)
- Use jarjar-gradle from github master i.e. '9a7eca7'(1.1.1); require gradle-8.2-bin.zip and precompiled smack jars
  * Gradle-8.x-bin.zip build without problem, but does not generate the smack jar files;
    Currently use this to build; and with previous generated smack jars;
  * Gradle-7.5-bin.zip generate smack jars in the defined directory, but not the defined name
    e.g. CH_1980125902.jar!!! need to find out how to use the generated smack jars files
- Update source to support API-34 ContextCompat.registerReceiver with RECEIVER_EXPORTED/RECEIVER_NOT_EXPORTED option.
- Update to 'com.android.tools.build:gradle:7.4.2'
- Update to 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.7.22'
- Upgrade Android Studio to Giraffe | 2022.3.1 - but with Gradle-JDK using jdk-15
- butterknife-compiler:10.2.3 is required for PermissionsActivity, but only compatible with jdk-15 max

=========================================================================
Version:        00.02.000 (0002000)
Release Date:   12/25/2023
Author:         cmeng
- Rename aTalk project namespace to org.atalk.ohos
- Strings/Strings_NT: cleanup and simplify string name id
- Cleanup resources.languages.resources.properties

=========================================================================
Version:        00.01.000 (0001000)
Release Date:   12/20/2023
Author:         cmeng
- Rebase org.atalk.hmos version to 00.01.000

/* === 12/12/2023 === */
- Add user selectable option 'Delete account on server' when user remove an account; user must login to perform this action.
- While user is login, changing local password will also change password on the server.
- Revert deletion of AudioSilenceSystem, AudioSilenceCaptureDevice, AudioSilenceMediaDevice classes
- Upgrade smack library to version 4.4.7
- Reset and cleanup mysql database version and creation.

/* === 11/28/2023 === */
- Remove SIP and ICQ protocols and preference settings support
- Remove aTalk unused codec, audio system and unused classes:
  * portaudio, pulseaudio, quicktime, and video4Linux protocol directory
  * PortAudioRenderer, PulseAudioRenderer, PortAudioSystem, PulseAudioSystem,
  * WASAPIRenderer, DirectShowSystem, QuickTimeSystem and Video4Linux2System
  * AudioSilenceSystem, AudioSilenceCaptureDevice, AudioSilenceMediaDevice,
- Build openssl jni library without installing the man document.

/* === 11/25/2023 === */
- Remove OTR encryption support
- Remove Desktop Util support
- Minimize use of android unsupported java classes e.g. java.jwt.Component, javax.swing.JComponent
- Remove unused classes, strings and drawables resources
- Update all aTalk JNI libraries build to use NDK '22.1.7171670';
  * Max NDK version that is compatible with aTalk JNI libraries built.
  * Retain ANDROID_API=21 for 64-bit and aTalk-android minSDK support;
  * Else built has errors e.g. ld: error: undefined symbol: stderr etc
- Upgrade jni libraries sources for:
  * libvpx 1.13.0 to 1.13.1; libvpx 1.3.1 source build requires NDK '22.1.7171670'
  * ffmpeg 5.1.2 to 5.1.4
- Clean up JNI scripts for all native libraries built
- Use the prebuilt toolchain instead of using make_standalone_toolchain.py.
- Update osmdroid-android:6.1.14 to 6.1.17

- In API-34: Sending file from recent items; the link provided is only temporary for the session
  and become invalid after the file transferred i.e. file deleted upon return to the session.

/* === 11/02/2023 === */
- Merged CallEnded class with VideoCallActivity
- Auto return to contactList 3 seconds after call has ended.

/* === 11/01/2023 === */
- Move project name space to org.atalk.hmos
- Change minimum API support to 24 (android-7)
- Migrate ice4j to only compatible with API-24


aTalk/src/main/java/org/atalk/hmos/gui/About.java
/*
 * aTalk, android VoIP and Instant Messaging client
 * Copyright 2014 Eng Chong Meng
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.atalk.hmos.gui;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.webkit.WebView;
import android.widget.TextView;

import net.java.sip.communicator.service.update.UpdateService;
import net.java.sip.communicator.util.ServiceUtils;

import org.atalk.hmos.BaseActivity;
import org.atalk.hmos.BuildConfig;
import org.atalk.hmos.R;
import org.atalk.hmos.aTalkApp;

import de.cketti.library.changelog.ChangeLog;

/**
 * About activity
 *
 * @author Eng Chong Meng
 */
public class About extends BaseActivity implements View.OnClickListener {
    private static final String[][] USED_LIBRARIES = new String[][]{
            new String[]{"Android Support Library", "https://developer.android.com/topic/libraries/support-library/index.html"},
            new String[]{"android-betterpickers", "https://github.com/code-troopers/android-betterpickers"},
            new String[]{"Android-EasyLocation", "https://github.com/akhgupta/Android-EasyLocation"},
            new String[]{"annotations-java5", "https://mvnrepository.com/artifact/org.jetbrains/annotations"},
            new String[]{"bouncycastle", "https://github.com/bcgit/bc-java"},
            new String[]{"butterknife", "https://github.com/JakeWharton/butterknife"},
            new String[]{"ckChangeLog", "https://github.com/cketti/ckChangeLog"},
            new String[]{"commons-lang", "https://commons.apache.org/proper/commons-lang/"},
            new String[]{"Dexter", "https://github.com/Karumi/Dexter"},
            new String[]{"dhcp4java", "https://github.com/ggrandes-clones/dhcp4java"},
            new String[]{"ExoPlayer", "https://github.com/google/ExoPlayer"},
            new String[]{"FFmpeg", "https://github.com/FFmpeg/FFmpeg"},
            new String[]{"glide", "https://github.com/bumptech/glide"},
            new String[]{"Google Play Services", "https://developers.google.com/android/guides/overview"},
            new String[]{"httpclient-android", "https://github.com/smarek/httpclient-android"},
            new String[]{"IPAddress", "https://github.com/seancfoley/IPAddress"},
            new String[]{"ice4j", "https://github.com/jitsi/ice4j"},
            new String[]{"jitsi", "https://github.com/jitsi/jitsi"},
            new String[]{"jitsi-android", "https://github.com/jitsi/jitsi-android"},
            new String[]{"jmdns", "https://github.com/jmdns/jmdns"},
            new String[]{"jxmpp-jid", "https://github.com/igniterealtime/jxmpp"},
            new String[]{"libjitsi", "https://github.com/jitsi/libjitsi"},
            new String[]{"libphonenumber", "https://github.com/googlei18n/libphonenumber"},
            new String[]{"libvpx", "https://github.com/webmproject/libvpx"},
            new String[]{"Mime4j", "https://james.apache.org/mime4j/"},
            new String[]{"miniDNS", "https://github.com/MiniDNS/minidns"},
            new String[]{"Noembed", "https://noembed.com/"},
            new String[]{"osmdroid", "https://github.com/osmdroid/osmdroid"},
            new String[]{"opensles", "https://github.com/openssl/openssl "},
            new String[]{"osgi.core", "http://grepcode.com/snapshot/repo1.maven.org/maven2/org.osgi/org.osgi.core/6.0.0"},
            new String[]{"sdes4j", "https://github.com/ibauersachs/sdes4j"},
            new String[]{"sdp-api", "https://mvnrepository.com/artifact/org.opentelecoms.sdp/sdp-api"},
            new String[]{"Smack", "https://github.com/igniterealtime/Smack"},
            new String[]{"speex", "https://github.com/xiph/speex"},
            new String[]{"Timber", "https://github.com/JakeWharton/timber"},
            new String[]{"TokenAutoComplete", "https://github.com/splitwise/TokenAutoComplete"},
            new String[]{"uCrop", "https://github.com/Yalantis/uCrop"},
            new String[]{"weupnp", "https://github.com/bitletorg/weupnp"},
            new String[]{"x264", "https://git.videolan.org/git/x264.git"},
            new String[]{"zrtp4j-light", "https://github.com/jitsi/zrtp4j"},
    };

    private static final String[][] SUPPORTED_XEP = new String[][]{
            new String[]{"XEP-0012: Last Activity 2.0", "https://xmpp.org/extensions/xep-0012.html"},
            new String[]{"XEP-0030: Service Discovery 2.5rc3", "https://xmpp.org/extensions/xep-0030.html"},
            new String[]{"XEP-0045: Multi-User Chat 1.34.3", "https://xmpp.org/extensions/xep-0045.html"},
            new String[]{"XEP-0047: In-Band Bytestreams 2.0.1", "https://xmpp.org/extensions/xep-0047.html"},
            new String[]{"XEP-0048: Bookmarks 1.2", "https://xmpp.org/extensions/xep-0048.html"},
            new String[]{"XEP-0054: vcard-temp 1.2", "https://xmpp.org/extensions/xep-0054.html"},
            new String[]{"XEP-0060: Publish-Subscribe 1.24.1", "https://xmpp.org/extensions/xep-0060.html"},
            new String[]{"XEP-0065: SOCKS5 Bytestreams 1.8.2", "https://xmpp.org/extensions/xep-0065.html"},
            new String[]{"XEP-0070: Verifying HTTP Requests via XMPP 1.0.1", "https://xmpp.org/extensions/xep-0070.html"},
            new String[]{"XEP-0071: XHTML-IM 1.5.4", "https://xmpp.org/extensions/xep-0071.html"},
            new String[]{"XEP-0077: In-Band Registration 2.4", "https://xmpp.org/extensions/xep-0077.html"},
            new String[]{"XEP-0084: User Avatar 1.1.4", "https://xmpp.org/extensions/xep-0084.html"},
            new String[]{"XEP-0085: Chat State Notifications 2.1", "https://xmpp.org/extensions/xep-0085.html"},
            new String[]{"XEP-0092: Software Version 1.1", "https://xmpp.org/extensions/xep-0092.html"},
            new String[]{"XEP-0095: Stream Initiation 1.2", "https://xmpp.org/extensions/xep-0095.html"},
            new String[]{"XEP-0096: SI File Transfer 1.3.1", "https://xmpp.org/extensions/xep-0096.html"},
            new String[]{"XEP-0100: Gateway Interaction 1.0", "https://xmpp.org/extensions/xep-0100.html"},
            new String[]{"XEP-0115: Entity Capabilities 1.6.0", "https://xmpp.org/extensions/xep-0115.html"},
            new String[]{"XEP-0124: Bidirectional-streams Over Synchronous HTTP (BOSH) 1.11.2", "https://xmpp.org/extensions/xep-0124.html"},
            new String[]{"XEP-0138: Stream Compression 2.1", "https://xmpp.org/extensions/xep-0138.html"},
            new String[]{"XEP-0153: vCard-Based Avatar 1.1", "https://xmpp.org/extensions/xep-0153.html"},
            new String[]{"XEP-0158: CAPTCHA Forms 1.5.8", "https://xmpp.org/extensions/xep-0158.html"},
            new String[]{"XEP-0163: Personal Eventing Protocol 1.2.2", "https://xmpp.org/extensions/xep-0163.html"},
            new String[]{"XEP-0166: Jingle 1.1.2", "https://xmpp.org/extensions/xep-0166.html"},
            new String[]{"XEP-0167: Jingle RTP Sessions 1.2.1", "https://xmpp.org/extensions/xep-0167.html"},
            new String[]{"XEP-0172: User Nickname 1.1", "https://xmpp.org/extensions/xep-0172.html"},
            new String[]{"XEP-0176: Jingle ICE-UDP Transport Method 1.1.1", "https://xmpp.org/extensions/xep-0176.html"},
            new String[]{"XEP-0177: Jingle Raw UDP Transport Method 1.1.1", "https://xmpp.org/extensions/xep-0177.html"},
            new String[]{"XEP-0178: Best Practices for Use of SASL EXTERNAL with Certificates 1.2", "https://xmpp.org/extensions/xep-0178.html"},
            new String[]{"XEP-0184: Message Delivery Receipts 1.4.0", "https://xmpp.org/extensions/xep-0184.html"},
            // new String[]{"XEP-0191: Blocking command (NI)", "https://xmpp.org/extensions/xep-0191.html"},
            new String[]{"XEP-0198: Stream Management 1.6", "https://xmpp.org/extensions/xep-0198.html"},
            new String[]{"XEP-0199: XMPP Ping 2.0.1", "https://xmpp.org/extensions/xep-0199.html"},
            new String[]{"XEP-0203: Delayed Delivery 2.0", "https://xmpp.org/extensions/xep-0203.html"},
            new String[]{"XEP-0206: XMPP Over BOSH 1.4", "https://xmpp.org/extensions/xep-0206.html"},
            new String[]{"XEP-0215: External Service Discovery 1.0.0", "https://xmpp.org/extensions/xep-0215.html"},
            new String[]{"XEP-0231: Bits of Binary 1.0", "https://xmpp.org/extensions/xep-0231.html"},
            new String[]{"XEP-0234: Jingle File Transfer 0.19.1", "https://xmpp.org/extensions/xep-0234.html"},
            new String[]{"XEP-0237: Roster Versioning 1.3", "https://xmpp.org/extensions/xep-0237.html"},
            new String[]{"XEP-0249: Direct MUC Invitations 1.2", "https://xmpp.org/extensions/xep-0249.html"},
            new String[]{"XEP-0251: Jingle Session Transfer 0.2", "https://xmpp.org/extensions/xep-0251.html"},
            new String[]{"XEP-0260: Jingle SOCKS5 Bytestreams Transport Method 1.0.3", "https://xmpp.org/extensions/xep-0260.html"},
            new String[]{"XEP-0261: Jingle In-Band Bytestreams Transport Method 1.0", "https://xmpp.org/extensions/xep-0261.html"},
            new String[]{"XEP-0262: Use of ZRTP in Jingle RTP Sessions 1.0", "https://xmpp.org/extensions/xep-0262.html"},
            new String[]{"XEP-0264: File Transfer Thumbnails 0.4", "https://xmpp.org/extensions/xep-0264.html"},
            new String[]{"XEP-0278: Jingle Relay Nodes 0.4.1", "https://xmpp.org/extensions/xep-0278.html"},
            new String[]{"XEP-0280: Message Carbons 1.0.1", "https://xmpp.org/extensions/xep-0280.html"},
            new String[]{"XEP-0293: Jingle RTP Feedback Negotiation 1.0.1", "https://xmpp.org/extensions/xep-0293.html"},
            new String[]{"XEP-0294: Jingle RTP Header Extensions Negotiation 1.1.1", "https://xmpp.org/extensions/xep-0294.html"},
            new String[]{"XEP-0298: Delivering Conference Information to Jingle Participants (Coin) 0.2", "https://xmpp.org/extensions/xep-0298.html"},
            new String[]{"XEP-0308: Last Message Correction 1.2.0", "https://xmpp.org/extensions/xep-0308.html"},
            new String[]{"XEP-0313: Message Archive Management 1.0.1", "https://xmpp.org/extensions/xep-0313.html"},
            new String[]{"XEP-0319: Last User Interaction in Presence 1.0.2", "https://xmpp.org/extensions/xep-0319.html"},
            new String[]{"XEP-0320: Use of DTLS-SRTP in Jingle Sessions 1.0.0", "https://xmpp.org/extensions/xep-0320.html"},
            new String[]{"XEP-0338: Jingle Grouping Framework 1.0.0", "https://xmpp.org/extensions/xep-0338.html"},
            new String[]{"XEP-0339: Source-Specific Media Attributes in Jingle 1.0.1", "https://xmpp.org/extensions/xep-0339.html"},
            new String[]{"XEP-0343: Signaling WebRTC datachannels in Jingle 0.3.1", "https://xmpp.org/extensions/xep-0343.html"},
            new String[]{"XEP-0352: Client State Indication 1.0.0", "https://xmpp.org/extensions/xep-0352.html"},
            new String[]{"XEP-0353: Jingle Message Initiation 0.4.0", "https://xmpp.org/extensions/xep-0353.html"},
            new String[]{"XEP-0363: HTTP File Upload 1.1.0", "https://xmpp.org/extensions/xep-0363.html"},
            new String[]{"XEP-0364: Off-the-Record Messaging (V2/3) 0.3.2", "https://xmpp.org/extensions/xep-0364.html"},
            new String[]{"XEP-0371: Jingle ICE Transport Method 0.3.1", "https://xmpp.org/extensions/xep-0371.html"},
            // new String[]{"XEP-0368: SRV records for XMPP over TLS", "https://xmpp.org/extensions/xep-0368.html"},
            new String[]{"XEP-0384: OMEMO Encryption 0.8.3", "https://xmpp.org/extensions/xep-0384.html"},
            new String[]{"XEP-0391: Jingle Encrypted Transports 0.1.2", "https://xmpp.org/extensions/xep-0391.html"},
            new String[]{"XEP-0441: Message Archive Management Preferences 0.2.0", "https://xmpp.org/extensions/xep-0441.htmll"},
            new String[]{"XEP-xxxx: OMEMO Media sharing 0.0.2", "https://xmpp.org/extensions/inbox/omemo-media-sharing.html"},
            // new String[]{"", ""},
    };

    /**
     * Default CSS styles used to format the change log.
     */
    public static final String DEFAULT_CSS =
            "h1 { margin-left: 0px; font-size: 1.2em; }" + "\n" +
                    "li { margin-left: 0px; font-size: 0.9em;}" + "\n" +
                    "ul { padding-left: 2em; }";

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.about);
        // crash if enabled under FragmentActivity
        // requestWindowFeature(Window.FEATURE_LEFT_ICON);
        // setFeatureDrawableResource(Window.FEATURE_LEFT_ICON, android.R.drawable.ic_dialog_info);
        setMainTitle(R.string.AboutDialog_title);

        View atakUrl = findViewById(R.id.atalk_link);
        atakUrl.setOnClickListener(this);

        TextView atalkHelp = findViewById(R.id.atalk_help);
        atalkHelp.setTextColor(getResources().getColor(R.color.blue50));
        atalkHelp.setOnClickListener(this);

        findViewById(R.id.ok_button).setOnClickListener(this);
        findViewById(R.id.history_log).setOnClickListener(this);

        View btn_submitLogs = findViewById(R.id.submit_logs);
        btn_submitLogs.setOnClickListener(this);

        if (BuildConfig.DEBUG) {
            View btn_update = findViewById(R.id.check_new_version);
            btn_update.setVisibility(View.VISIBLE);
            btn_update.setOnClickListener(this);
        }

        String aboutInfo = getAboutInfo();
        WebView wv = findViewById(R.id.AboutDialog_Info);
        wv.loadDataWithBaseURL("file:///android_res/drawable/", aboutInfo, "text/html", "utf-8", null);

        try {
            PackageInfo pi = getPackageManager().getPackageInfo(getPackageName(), 0);

            TextView textView = findViewById(R.id.AboutDialog_Version);
            textView.setText(String.format(getString(R.string.AboutDialog_Version), pi.versionName));
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.ok_button:
                finish();
                break;
            case R.id.check_new_version:
                new Thread() {
                    @Override
                    public void run() {
                        UpdateService updateService
                                = ServiceUtils.getService(AndroidGUIActivator.bundleContext, UpdateService.class);
                        if (updateService != null) {
                            updateService.checkForUpdates();
                        }
                    }
                }.start();
                break;
            case R.id.submit_logs:
                aTalkApp.showSendLogsDialog();
                break;
            case R.id.history_log:
                ChangeLog cl = new ChangeLog(this, DEFAULT_CSS);
                cl.getFullLogDialog().show();
                break;
            case R.id.atalk_help:
            case R.id.atalk_link:
                atalkUrlAccess(this, getString(R.string.AboutDialog_Link));
                break;
            default:
                finish();
                break;
        }
    }

    public static void atalkUrlAccess(Context context, String url) {
        if (url == null)
            url = context.getString(R.string.AboutDialog_Link);
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(url));
        context.startActivity(intent);
    }

    private String getAboutInfo() {
        StringBuilder html = new StringBuilder()
                .append("<meta http-equiv=\"content-type\" content=\"text/html; charset=utf-8\"/>")
                .append("<html><head><style type=\"text/css\">")
                .append(DEFAULT_CSS)
                .append("</style></head><body>");


        StringBuilder xeps = new StringBuilder().append("<ul>");
        for (String[] feature : SUPPORTED_XEP) {
            xeps.append("<li><a href=\"")
                    .append(feature[1])
                    .append("\">")
                    .append(feature[0])
                    .append("</a></li>");
        }
        xeps.append("</ul>");

        html.append(String.format(getString(R.string.app_xeps), xeps.toString()))
                .append("</p><hr/><p>");

        StringBuilder libs = new StringBuilder().append("<ul>");
        for (String[] library : USED_LIBRARIES) {
            libs.append("<li><a href=\"")
                    .append(library[1])
                    .append("\">")
                    .append(library[0])
                    .append("</a></li>");
        }
        libs.append("</ul>");

        html.append(String.format(getString(R.string.app_libraries), libs.toString()))
                .append("</p><hr/><p>");
        html.append("</body></html>");

        return html.toString();
    }
}
